### PR TITLE
Fitness rebalance and improvements - mood, fireman carry, and longer workouts

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request
So after getting feedback on fitness I've decided to make some balance changes:

- ~~Every level of fitness now increases max hp. At max legendary fitness, all the hp bonuses combined results in a total of +25 max hp~~
- Fitness now decreases the time it takes to firemany carry by 0.33 seconds per level. (At max fitness, this means a decrease of 2 seconds)
- Exercise status effect triggers the exercise mood event and goes away when the exercise status effect is gone
- The better your fitness level, the more happiness you gain from the mood event
- Decreased the fitness timers and effects by half and tweaked a few other values
- Increased the nutrition cost of working out
- Removed doubles metabolism exercise effect

I'd also like this test merged to see if anymore minor tweaks need to be made.

## Why It's Good For The Game
Fitness effects on gameplay were pretty underwhelming.  This was deliberate since there is a concern about prisoners being able to abuse it in quite a lot of scenarios.  ~~Adding a small boost to max hp is nice given that the downside is a bigger target.~~

Another problem was that the double metabolism rate resulted in a lot of the diet effects being metabolised before you could convert them into exercise gains.  So this was removed.

People would also hit their bench max rather quickly and then spend a long duration sleeping.  Ideally I wanted people to spend more time working out than sleeping, so I halved the duration values so it takes them longer to achieve maxxing out.

## Changelog
:cl:
balance: Fitness level decreases the time it takes to firemany carry someone. Fitness level determines how much of a positive mood the workout grants. Working out is now more difficult and requires more nutrition.
balance: Exercise no longer triggers double metabolism.  
/:cl:

### CODE REVIEW

**Comment by MrMelbert** [comment]
It's worth noting increasing human max health doesn't really *work* as you think, in fact it barely works at all

File: code/datums/skills/fitness.dm [hunk 1]
```diff
@@ -3,17 +3,24 @@
...
 	// skill_item_path - your mob sprite gets bigger to showoff so we don't get a special item
 
 /datum/skill/fitness/level_gained(datum/mind/mind, new_level, old_level, silent)
 	. = ..()
 	var/size_boost = (new_level == SKILL_LEVEL_LEGENDARY) ? 0.25 : 0.05
 	var/gym_size = RESIZE_DEFAULT_SIZE + size_boost
 	mind.current.update_transform(gym_size)
+	// at max legendary fitness, all the hp bonuses combined results in a total of +25 max hp
+	mind.current.maxHealth += (size_boost*100)/2 // +2.5 hp per level (and +5 hp on gaining legendary)
+	mind.current.health += (size_boost*100)/2
```
Comment (Fikou) [inline code/datums/skills/fitness.dm:16]: i dont think youre allowed to do that

**Comment by timothymtorres** [comment]
Tested and it works fine.

dreamseeker_gBrGjad6Dj

Whatever you set as the max health is the threshold when a mob goes into crit.  So if you set it to 200, the mob can take 200 damage before crit happens.

File: code/datums/skills/fitness.dm (hunk 1 above)
Comment (timothymtorres) [reply code/datums/skills/fitness.dm:16]: I tested it and works fine.  There isn't a proc to directly adjust maxhealth and the other parts of the codebase that modify it do it directly like this.  Do you need me to refactor this and the other parts into an `adjustMaxHealth` proc or something?

**Comment by atlasle** [comment]
shouldn't it give you damage resistance instead of max health, max health kinda makes it sound like you're getting more flesh on you

**Comment by AtomTheProphet** [comment]
> shouldn't it give you damage resistance instead of max health, max health kinda makes it sound like you're getting more flesh on you

I mean.. that is kind of what the point of exercise sometimes is, is it not? To develop muscles and whatnot.

**Comment by ZephyrTFA** [comment]
Don't really like increasing max health here. Your health is calculated based on damage to your limbs, which have a max damage amount.
If we buff max health is incredibly likely we get to the point where you literally cannot put someone into crit

**Comment by moocowswag** [comment]
> Don't really like increasing max health here. Your health is calculated based on damage to your limbs, which have a max damage amount. If we buff max health it is likely we get to the point in the future where you literally cannot put someone into crit
crit is still possible i think based on my memory from watermelons rng stat mechanic that admins kept implimenting.

but it sucked so fucking bad when an assistant had 400 health and strong punches that nobody else had so he was tanking several stun baton hits and still easily stamcritting others :(

**Comment by moocowswag** [comment]
Okay so one major effect of this pr, im like 80% sure that just the first level of fitness combined with decent mood will mean you take 3 stun baton hits to go into stamcrit instead of two.

**Comment by timothymtorres** [comment]
> Okay so one major effect of this pr, im like 80% sure that just the first level of fitness combined with decent mood will mean you take 3 stun baton hits to go into stamcrit instead of two.
> 
> 
> 
> 
> 
> actually thats probably THE most impactful because it means people WILL 100% do the bare minimum of working out a tiny bit before going to fight security.

This PR doesn't touch stamina other than using it to workout.

**Comment by ZephyrTFA** [comment]
> > Okay so one major effect of this pr, im like 80% sure that just the first level of fitness combined with decent mood will mean you take 3 stun baton hits to go into stamcrit instead of two.
> > actually thats probably THE most impactful because it means people WILL 100% do the bare minimum of working out a tiny bit before going to fight security.
> 
> This PR doesn't touch stamina other than using it to workout.

The effect of stamina damage is based on your max health

**Comment by moocowswag** [comment]
a stun baton does 60 stamina damage, 2 hits is 120 which is exactly the amount a max mood spaceman has.

either it triggers the funny stamina regen nerf code that sets your stamina to 150 on stamcrit or it doesnt and you get another free hit.

**Comment by timothymtorres** [comment]
So it appears stamcrit is affected by maxhealth.

https://github.com/tgstation/tgstation/blob/f9a5502f78263690378ca01f660b1cfb2886efd1/code/modules/mob/living/carbon/carbon.dm#L512-L516

Going to look for a good solution to avoid this.  Also I did give a boost to firemany carrying so that it shaves off 1/3 of a second for each level of fitness.

File: code/datums/skills/fitness.dm (hunk 1 above)
Comment (timothymtorres) [reply code/datums/skills/fitness.dm]: Refactored this to be more readable.

File: code/game/objects/structures/gym/punching_bag.dm [hunk 2]
```diff
@@ -51,10 +51,10 @@
...
 			stamina_exhaustion = 2
 
-	user.adjustStaminaLoss(stamina_exhaustion)
-	user.add_mood_event("exercise", /datum/mood_event/exercise)
-	user.mind?.adjust_experience(/datum/skill/fitness, 0.1)
-	user.apply_status_effect(/datum/status_effect/exercised)
+	if(iscarbon(user))
+		user.adjustStaminaLoss(stamina_exhaustion)
+		user.mind?.adjust_experience(/datum/skill/fitness, 0.1)
+		user.apply_status_effect(/datum/status_effect/exercised)
```
Comment (ZephyrTFA) [inline code/game/objects/structures/gym/punching_bag.dm:57]:
```suggestion
	if(!iscarbon(user))
		return
	user.adjustStaminaLoss(stamina_exhaustion)
	user.mind?.adjust_experience(/datum/skill/fitness, 0.1)
	user.apply_status_effect(/datum/status_effect/exercised)
```

File: code/game/objects/structures/gym/weight_machine.dm [hunk 3]
```diff
@@ -150,15 +156,21 @@
 	flick_overlay_view(workout, 0.8 SECONDS)
 	flick("[base_icon_state]-u", src)
 	var/mob/living/user = buckled_mobs[1]
-	animate(user, pixel_y = pixel_shift_y, time = 4)
+	animate(user, pixel_y = pixel_shift_y, time = WORKOUT_LENGTH/2)
```
Comment (ZephyrTFA) [inline code/game/objects/structures/gym/weight_machine.dm:159]:
```suggestion
	animate(user, pixel_y = pixel_shift_y, time = WORKOUT_LENGTH * 0.5)
```

File: code/game/objects/structures/gym/weight_machine.dm [hunk 4]
```diff
@@ -150,15 +156,21 @@
...
 	playsound(user, 'sound/machines/creak.ogg', 60, TRUE)
-	animate(pixel_y = user.base_pixel_y, time = 4)
+	animate(pixel_y = user.base_pixel_y, time = WORKOUT_LENGTH/2)
```
Comment (ZephyrTFA) [inline code/game/objects/structures/gym/weight_machine.dm:161]:
```suggestion
	animate(pixel_y = user.base_pixel_y, time = WORKOUT_LENGTH * 0.5)
```

File: code/game/objects/structures/gym/weight_machine.dm [hunk 5]
```diff
@@ -150,15 +156,21 @@
...
 
-	var/stamina_exhaustion = 5 - (user.mind.get_skill_level(/datum/skill/fitness) * 0.5)
-	user.adjustStaminaLoss(stamina_exhaustion * seconds_per_tick)
+	if(iscarbon(user) && user.mind)
+		// the amount of workouts you can do before you hit stamcrit
+		var/workout_reps = total_workout_reps[user.mind.get_skill_level(/datum/skill/fitness)]
+		// total stamina drain of 1 workout calculated based on the workout length
+		var/stamina_exhaustion = FLOOR(initial(user.maxHealth) / workout_reps / WORKOUT_LENGTH, 0.1)
+		user.adjustStaminaLoss(stamina_exhaustion * seconds_per_tick)
```
Comment (ZephyrTFA) [inline code/game/objects/structures/gym/weight_machine.dm:168]:
```suggestion
	if(!iscarbon(user) || isnull(user.mind))
		return TRUE
	// the amount of workouts you can do before you hit stamcrit
	var/workout_reps = total_workout_reps[user.mind.get_skill_level(/datum/skill/fitness)]
	// total stamina drain of 1 workout calculated based on the workout length
	var/stamina_exhaustion = FLOOR(initial(user.maxHealth) / workout_reps / WORKOUT_LENGTH, 0.1)
	user.adjustStaminaLoss(stamina_exhaustion * seconds_per_tick)
```

File: code/modules/mob/living/carbon/carbon.dm [hunk 6]
```diff
@@ -530,7 +530,7 @@
 
 /mob/living/carbon/update_stamina()
 	var/stam = getStaminaLoss()
-	if(stam > DAMAGE_PRECISION && (maxHealth - stam) <= crit_threshold)
+	if(stam > DAMAGE_PRECISION && (initial(maxHealth) - stam) <= crit_threshold)
```
Comment (ZephyrTFA) [inline code/modules/mob/living/carbon/carbon.dm:533]: should probably be documented this works off initial and not current

File: code/modules/mob/living/carbon/human/human.dm [hunk 7]
```diff
@@ -869,11 +869,18 @@
 
 	var/carrydelay = 5 SECONDS //if you have latex you are faster at grabbing
 	var/skills_space = "" //cobby told me to do this
+	var/fitness_level = src.mind.get_skill_level(/datum/skill/fitness) - 1
```
Comment (ZephyrTFA) [inline code/modules/mob/living/carbon/human/human.dm:872]:
```suggestion
	var/fitness_level = mind.get_skill_level(/datum/skill/fitness) - 1
```

File: code/modules/mob/living/carbon/human/human.dm [hunk 8]
```diff
@@ -869,11 +869,18 @@
 
 	var/carrydelay = 5 SECONDS //if you have latex you are faster at grabbing
 	var/skills_space = "" //cobby told me to do this
```
Comment (ZephyrTFA) [inline code/modules/mob/living/carbon/human/human.dm:871]:
```suggestion
	var/skills_space
```

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/carbon.dm (hunk 6 above)
Comment (timothymtorres) [reply code/modules/mob/living/carbon/carbon.dm:534]: Applied with https://github.com/tgstation/tgstation/pull/79584/commits/247791f452f3f57f922feacf6aaab663144f2f85

**Review by Fikou — CHANGES_REQUESTED** [review]
still not a fan of modifying your max health

**Comment by Mothblocks** [comment]
Drafted pending FIkou's request

File: code/datums/skills/fitness.dm [hunk 9]
```diff
@@ -3,17 +3,25 @@
...
 	. = ..()
-	var/size_boost = (new_level == SKILL_LEVEL_LEGENDARY) ? 0.25 : 0.05
-	var/gym_size = RESIZE_DEFAULT_SIZE + size_boost
-	mind.current.update_transform(gym_size)
+	var/old_gym_size = RESIZE_DEFAULT_SIZE + size_boost[old_level]
+	var/new_gym_size = RESIZE_DEFAULT_SIZE + size_boost[new_level]
+
+	// reset the size before applying the new one, otherwise the size boosts will stack
+	mind.current.update_transform(RESIZE_DEFAULT_SIZE / old_gym_size)
+	mind.current.update_transform(new_gym_size)
```
Comment (Ghommie) [inline code/datums/skills/fitness.dm:18]:
```suggestion
	mind.current.update_transform(new_gym_size / old_gym_size)
```

File: code/datums/skills/fitness.dm [hunk 10]
```diff
@@ -3,17 +3,25 @@
...
 	. = ..()
-	var/size_boost = (new_level == SKILL_LEVEL_LEGENDARY) ? 0.25 : 0.05
-	var/gym_size = RESIZE_DEFAULT_SIZE + size_boost
-	mind.current.update_transform(RESIZE_DEFAULT_SIZE / gym_size)
+	var/old_gym_size = RESIZE_DEFAULT_SIZE + size_boost[old_level]
+	var/new_gym_size = RESIZE_DEFAULT_SIZE + size_boost[new_level]
+
+	// reset the size before applying the new one, otherwise the size boosts will stack
+	mind.current.update_transform(RESIZE_DEFAULT_SIZE / old_gym_size)
+	mind.current.update_transform(new_gym_size)
```
Comment (Ghommie) [inline code/datums/skills/fitness.dm:27]:
```suggestion
	mind.current.update_transform(new_gym_size / old_gym_size)
```

**Review by Ghommie — CHANGES_REQUESTED** [review]

**Comment by Ghommie** [comment]
Merge conflicts. Once resolved, I'll go ahead and merge the PR, now that Fikou's request has gone through.

**Pull request merged**